### PR TITLE
[ie/Facebook] Fix Unhandled IndexError

### LINKED ISSUE #8669: Unhandled IndexError when downloading video from Facebook
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I *intentionally* remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting that yt-dlp is broken on a **supported** site
- [X] I've verified that I have **updated yt-dlp to nightly or master** ([update instructions](https://github.com/yt-dlp/yt-dlp#update-channels))
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/yt-dlp/yt-dlp/wiki/FAQ#video-url-contains-an-ampersand--and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched [known issues](https://github.com/yt-dlp/yt-dlp/issues/3766) and the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [ ] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

USA

### Provide a description that is worded well enough to be understood

Action: Attempting to download a video from a Facebook URL.  The video is publicly viewable and plays normally in my browser (firefox) without logging in.

Expected behavior: the video is downloaded to the current working directory.

Actual behavior: I am met with `IndexError: list index out of range`.

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] If using API, add `'verbose': True` to `YoutubeDL` params instead
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
$ ./yt-dlp_linux --ignore-config -vU "https://www.facebook.com/groups/1645456212344334/permalink/3737828833107051/?ref=share&mibextid=NSMWBT"
[debug] Command-line config: ['--ignore-config', '-vU', 'https://www.facebook.com/groups/1645456212344334/permalink/3737828833107051/?ref=share&mibextid=NSMWBT']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version nightly@2023.11.26.232703 from yt-dlp/yt-dlp-nightly-builds [9751a457c] (linux_exe)
[debug] Python 3.10.13 (CPython x86_64 64bit) - Linux-6.6.2-zen1-1-zen-x86_64-with-glibc2.38 (OpenSSL 3.2.0 23 Nov 2023, glibc 2.38)
[debug] exe versions: ffmpeg 6.1 (setts), ffprobe 6.1, rtmpdump 2.4
[debug] Optional libraries: Cryptodome-3.19.0, brotli-1.1.0, certifi-2023.11.17, mutagen-1.47.0, requests-2.31.0, secretstorage-3.3.3, sqlite3-3.44.2, urllib3-2.1.0, websockets-12.0
[debug] Proxy map: {}
[debug] Request Handlers: urllib, requests, websockets
[debug] Loaded 1792 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp-nightly-builds/releases/latest
Latest version: nightly@2023.11.26.232703 from yt-dlp/yt-dlp-nightly-builds
yt-dlp is up to date (nightly@2023.11.26.232703 from yt-dlp/yt-dlp-nightly-builds)
[facebook] Extracting URL: https://www.facebook.com/groups/1645456212344334/permalink/3737828833107051/?ref=share&mibextid=NSMWBT
[facebook] 3737828833107051: Downloading webpage
ERROR: list index out of range
Traceback (most recent call last):
  File "yt_dlp/YoutubeDL.py", line 1570, in wrapper
  File "yt_dlp/YoutubeDL.py", line 1705, in __extract_info
  File "yt_dlp/extractor/common.py", line 717, in extract
  File "yt_dlp/extractor/facebook.py", line 778, in _real_extract
  File "yt_dlp/extractor/facebook.py", line 640, in _extract_from_url
IndexError: list index out of range
```

### PR DESCRIPTION
**Pull request by kclauhk**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

ADD DESCRIPTION HERE

Fixes #8669 

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

### CODE REVIEW

File: yt_dlp/extractor/facebook.py [hunk 1]
```diff
@@ -637,7 +638,7 @@ def parse_attachment(attachment, key='media'):
                 if len(entries) > 1:
                     return self.playlist_result(entries, video_id)
 
-                video_info = entries[0]
+                video_info = entries[0] if len(entries) > 0 else {'id': video_id}
```
Comment (pukkandan) [inline yt_dlp/extractor/facebook.py:641]: Could you explain this change?

Comment (pukkandan) [reply yt_dlp/extractor/facebook.py:641]:
```suggestion
                video_info = entries[0] entries else {'id': video_id}
```

File: yt_dlp/extractor/facebook.py [hunk 2]
```diff
@@ -612,9 +612,10 @@ def parse_attachment(attachment, key='media'):
                 nodes = variadic(traverse_obj(data, 'nodes', 'node') or [])
                 attachments = traverse_obj(nodes, (
                     ..., 'comet_sections', 'content', 'story', (None, 'attached_story'), 'attachments',
-                    ..., ('styles', 'style_type_renderer'), 'attachment'), expected_type=dict) or []
+                    ..., ('styles', 'throwbackStyles', 'style_type_renderer'), (None, ...), 'attachment'), expected_type=dict) or []
                 for attachment in attachments:
-                    ns = try_get(attachment, lambda x: x['all_subattachments']['nodes'], list) or []
+                    ns = (try_get(attachment, lambda x: x['all_subattachments']['nodes'], list)
+                          or traverse_obj(attachment, (..., 'attachments', ..., 'styles', 'attachment'), expected_type=dict) or [])
```
Comment (pukkandan) [inline yt_dlp/extractor/facebook.py:618]:
```suggestion
                    ns = traverse_obj(attachment, ('all_subattachments', 'nodes', ..., {dict}),
                                      (..., 'attachments', ..., 'styles', 'attachment', {dict}))
```

File: yt_dlp/extractor/facebook.py (hunk 1 above)
Comment (kclauhk) [reply yt_dlp/extractor/facebook.py]: without this change, "list index out of range" still occurs if line 613-615 cannot find anything.
with this change, "No format found" will be thrown.

**Review by seproDev — CHANGES_REQUESTED** [review]
Could you also please add a test case for a video linked to from a memory?

File: yt_dlp/extractor/facebook.py [hunk 3]
```diff
@@ -232,6 +232,19 @@ class FacebookIE(InfoExtractor):
...
+        'url': 'https://www.facebook.com/groups/1645456212344334/posts/3737828833107051/',
+        'info_dict': {
+            'id': '1569199726448814',
+            'ext': 'mp4',
+            'title': 'Pence MUST GO!',
+            'description': 'Vickie Gentry shared a memory.',
+            'timestamp': 1511548260,
+            'upload_date': '20171124',
+            'uploader': 'Vickie Gentry',
+            'uploader_id': 'pfbid0FuZhHCeWDAxWxEbr3yKPFaRstXvRxgsp9uCPG6GjD4J2AitB35NUAuJ4Q75KcjiDl',
```
Comment (seproDev) [inline yt_dlp/extractor/facebook.py:246]: The test is incomplete.
```suggestion
            'uploader_id': 'pfbid0FuZhHCeWDAxWxEbr3yKPFaRstXvRxgsp9uCPG6GjD4J2AitB35NUAuJ4Q75KcjiDl',
            'thumbnail': r're:^https?://.*',
            'duration': 148.435,
```

File: yt_dlp/extractor/facebook.py [hunk 4]
```diff
@@ -612,9 +627,11 @@ def parse_attachment(attachment, key='media'):
                 nodes = variadic(traverse_obj(data, 'nodes', 'node') or [])
                 attachments = traverse_obj(nodes, (
                     ..., 'comet_sections', 'content', 'story', (None, 'attached_story'), 'attachments',
-                    ..., ('styles', 'style_type_renderer'), 'attachment'), expected_type=dict) or []
+                    ..., ('styles', 'style_type_renderer', ('throwbackStyles', 'attachment_target_renderer')),
+                    'attachment'), expected_type=dict) or []
```
Comment (bashonly) [inline yt_dlp/extractor/facebook.py:631]: we're branching, so `or []` is unnecessary
```suggestion
                    ..., ('styles', 'style_type_renderer', ('throwbackStyles', 'attachment_target_renderer')),
                    'attachment', {dict}))
```

**Review by bashonly — APPROVED** [review]
note for maintainers: consider that the scope of this PR has expanded beyond just handling the indexerror when writing the commit msg

**Pull request merged**